doc/adr/0005-jwt-authentication.md

```diff
@@ -0,0 +1,157 @@
+# 5. JWT Authentication Extensibility
+
+Date: 2020-07-01
+
+## Status
+
+Proposed.
+
+
+## Context
+
+We wish to have the option to, in future, extend or change the authentication
+flow in a way which clients can safely handle.  This architecture design note
+provides for a check which clients can safely add now, to be compatible with
+such future changes and to be proof against certain attacks.
+
+Upon connection, the NATS server sends an `INFO` line with various fields; the
+NATS client processes that, performs a TLS upgrade if appropriate, and then
+sends a `CONNECT` line with various fields.
+
+At present, if a server supports JWT authentication then it generates a strong
+random nonce, presented as the `nonce` field of the `INFO` line.  If a client
+wishes to authenticate with a JWT, then it takes the bytes of that field, and
+without any decoding signs them with the private key associated with the jwt;
+the resulting signature is base64-encoded; then in the `CONNECT` line the
+client includes two fields: `jwt` containing their JWT, and `sig` which
+contains the base64-encoded result of signing the nonce.
+
+This protects the server against replay attacks, since a strongly random nonce
+is chosen each time.  It provides no inherent protection against
+man-in-the-middle attacks, relying instead upon TLS checks.  The server can
+enforce mutual-TLS, but not all deployment scenarios encoforce this.
+
+A well-behaved client with a competent TLS stack, configured to only proceed
+with TLS, will verify the NATS server's certificate before proceeding, so will
+only send the signature to a server which can present a certificate that
+client expects.
+
+In the IoT space, and in some programming languages, the TLS libraries don't
+necessarily actually perform full certificate verification.  While there are
+limits to how much "reinventing the wheel" should be done here, there is
+established precedence in authentication protocols for being proof against
+such scenarios.
+
+One notable caveat here is the "enterprise SSL inspection" man-in-the-middle,
+which presents a certificate which the client expects.  In practice, NATS will
+not get through any such device which does not explicitly support NATS,
+because such boxes primarily target TLS-on-connect and then add
+protocol-specific workarounds for all other scenarios.
```

Comment:
One exception is Websocket clients for which we do TLS handshake first.
I also think that we should have the option in both server and client to do TLS right away.

Reply:
Oops; reworded to handle that I'm just describing the core flow.  Agreed re TLS-on-connect as a nice thing, but here I'm just setting the context to help people envisage the data flows.  I don't think TLS-on-connect changes the JWT signing here at all.